File: src/bot/security/slowMode.js
import { ChannelType, PermissionsBitField } from "discord.js";
import { getConfig } from "../../db/database.js";

export class SlowModeManager {
  constructor(logger) {
    this.logger = logger;
    this.messageBuckets = new Map();
    this.lastApplied = new Map();
    this.lastChange = new Map();
  }

  handleMessage(message) {
    if (!message.guild || message.author.bot) return;

    const config = getConfig();
    const slowModeConfig = config.slowMode;
    if (!slowModeConfig?.enabled) return;

    const channel = message.channel;
    if (!this.canApplySlowMode(channel, message.guild)) return;

    const now = Date.now();
    const windowMs = (slowModeConfig.windowSeconds || 60) * 1000;
    const bucket = this.messageBuckets.get(channel.id) || [];
    const fresh = bucket.filter((ts) => now - ts < windowMs);
    fresh.push(now);
    this.messageBuckets.set(channel.id, fresh);

    const ratePerMinute = (fresh.length * 60000) / windowMs;
    const target = this.selectSlowMode(ratePerMinute, slowModeConfig.tiers || []);

    const current = this.lastApplied.get(channel.id) ?? channel.rateLimitPerUser ?? 0;
    if (target === current) return;

    const minInterval = (slowModeConfig.minUpdateIntervalSeconds || 15) * 1000;
    const lastChange = this.lastChange.get(channel.id) || 0;
    const isIncrease = target > current;
    if (!isIncrease && now - lastChange < minInterval) return;

    channel
      .setRateLimitPerUser(target, "Automatic slow mode adjustment")
      .then(() => {
        this.lastApplied.set(channel.id, target);
        this.lastChange.set(channel.id, now);
        const action = target === 0 ? "disabled" : `${target}s`;
        this.logger(channel.guild.id, `⏱️ Slow mode ${action} in #${channel.name} (${ratePerMinute.toFixed(1)} msg/min).`);
      })
      .catch(() => {});
  }

  canApplySlowMode(channel, guild) {
    const allowedTypes = [ChannelType.GuildText, ChannelType.GuildAnnouncement];
    if (!channel || !allowedTypes.includes(channel.type)) return false;
    if (typeof channel.setRateLimitPerUser !== "function") return false;
    const me = guild.members.me;
    if (!me) return false;
    const permissions = channel.permissionsFor(me);
    return permissions?.has(PermissionsBitField.Flags.ManageChannels);
  }

  selectSlowMode(ratePerMinute, tiers) {
    let target = 0;
    for (const tier of tiers) {
      if (ratePerMinute >= tier.threshold) {
        target = tier.seconds;
      }
    }
    return target;
  }
}

File: bot/slow_mode.py
import datetime
from collections import defaultdict, deque
from typing import Any

import discord


class SlowModeManager:
    def __init__(self, bot: discord.Client, config: dict[str, Any]):
        self.bot = bot
        self.message_buckets: dict[int, deque] = defaultdict(lambda: deque(maxlen=1000))
        self.last_applied: dict[int, int] = {}
        self.last_change: dict[int, datetime.datetime] = {}
        self.update_config(config)

    def update_config(self, config: dict[str, Any]) -> None:
        self.config = (config or {}).get('slow_mode') or {}

    def handle_message(self, message: discord.Message) -> None:
        if message.guild is None or message.author.bot:
            return
        conf = self.config or {}
        if not conf.get('enabled', False):
            return
        now = datetime.datetime.utcnow()
        bucket = self.message_buckets[message.channel.id]
        bucket.append(now)
        window = max(int(conf.get('window_seconds', 60)), 10)
        while bucket and (now - bucket[0]).total_seconds() > window:
            bucket.popleft()
        rate = (len(bucket) / window) * 60
        tiers = conf.get('tiers') or []
        target = self._select_slowmode(rate, tiers)
        last = self.last_applied.get(message.channel.id, 0)
        cooldown = max(int(conf.get('min_update_interval_seconds', 15)), 5)
        if target == last:
            return
        last_change = self.last_change.get(message.channel.id)
        if last_change and (now - last_change).total_seconds() < cooldown:
            return
        if isinstance(message.channel, discord.TextChannel):
            try:
                self.bot.loop.create_task(message.channel.edit(slowmode_delay=target, reason='Auto slow mode'))
                self.last_applied[message.channel.id] = target
                self.last_change[message.channel.id] = now
            except Exception:
                return

    def _select_slowmode(self, rate_per_minute: float, tiers: list[dict[str, Any]]) -> int:
        selected = 0
        for tier in sorted(tiers, key=lambda t: t.get('threshold', 0), reverse=True):
            try:
                threshold = float(tier.get('threshold', 0))
                seconds = int(tier.get('seconds', 0))
            except (TypeError, ValueError):
                continue
            if rate_per_minute >= threshold:
                selected = max(selected, max(seconds, 0))
                break
        return selected
